Adding function state change traces to improve orchestration monitoring query

bachuv opened the pull request:
[Creating this as a draft PR for now to make sure all tests pass]

This PR adds new traces to track function state changes. This helps us improve our monitoring query in the portal (Functions -> OrchestratorFunction -> Monitor tab -> App Insights query).

New query to list the status of orchestration instances:
```
traces
| where customDimensions.Category == "Host.Triggers.DurableTask"
| extend functionName = tostring(customDimensions["prop__functionName"])
| extend instanceId = tostring(customDimensions["prop__instanceId"])
| extend state = tostring(customDimensions["prop__functionState"])
| extend isReplay = tobool(tolower(customDimensions["prop__isReplay"]))
| where isReplay != true
| where functionName == <functionName>
| where state != ""
| summarize arg_max(timestamp, *) by instanceId
| project timestamp, instanceId, state
```


### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)
* [ ] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [ ] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`

cgillum commented on src/WebJobs.Extensions.DurableTask/Listener/TaskEntityShim.cs:
```diff
@@ -170,6 +170,14 @@ internal void Rehydrate(string serializedInput)
+                    this.context.Name,
+                    this.context.InstanceId,
+                    FunctionType.Orchestrator,
+                    FunctionState.Failed,
```
Why is this considered a failure? I'm not super familiar with the entity code, but line 168 gives the impression that the entity is actually being *created*.

davidmrdavid replied:
Relatedly - should this have `FunctionType.Entity` as a parameter instead?

davidmrdavid commented on src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs:
```diff
@@ -393,6 +393,14 @@ async Task IDurableOrchestrationClient.TerminateAsync(string instanceId, string
                 await this.client.TerminateInstanceAsync(state.OrchestrationInstance, reason);
 
                 this.traceHelper.FunctionTerminated(this.TaskHubName, state.Name, instanceId, reason);
+
+                this.traceHelper.OrchestratorStateChange(
+                    this.TaskHubName,
+                    state.Name,
+                    instanceId,
+                    FunctionType.Orchestrator,
+                    FunctionState.Terminated,
+                    isReplay: false);
             }
```
Have we tested, on Azure, that terminating an orchestrator mid-execution won't lead it to accidentally emit a "Terminated" state-change log **and then** a "Completed" state-change log?

This was happening to us when when the following events took place:
1. orchestrator starts and schedules a 1 minute timer
2. before the 1 minute expires, the orchestrator is terminated. Therefore, the terminated state-change log is emitted
3. the time fires, and the orchestrator is resumed
4. the orchestrator performs a replay as a result of the timer firing. Therefore, it emits a completed state-change log.

I would appreciate if we could perform this test. It would help me gain confidence in the approach :)

davidmrdavid reviewed (commented):
Thanks for doing this work! I added a request for a manual test, please see my comment below

davidmrdavid commented on src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs:
```diff
@@ -962,6 +962,23 @@ public void ExtensionWarningEvent(string hubName, string functionName, string in
+            if (this.ShouldLogEvent(isReplay))
+            {
+                this.logger.LogInformation(
+                    "{instanceId}: Function '{functionName} ({functionType})' state changed to {functionState}. IsReplay: {isReplay}. State: {state}. FunctionState: {functionState}. HubName: {hubName}. AppName: {appName}. SlotName: {slotName}. ExtensionVersion: {extensionVersion}. SequenceNumber: {sequenceNumber}.",
+                    instanceId, functionName, functionType, functionState, isReplay, functionState, functionState, hubName,
+                    LocalAppName, LocalSlotName, ExtensionVersion, this.sequenceNumber++);
```
It seems we're logging `functionState` twice, aren't we? Once for `State: ` and another time for `FunctionState: `.

Should we just log it in the `State: ` field?

bachuv commented:
As an update, I removed the `functionState` property to make it simpler and updated the query in the description based on that change. I'm now looking into removing the additional "orchestrator state changed" log.

bachuv commented:
I looked into removing the additional state trace log and updating the existing `Activity.Tags` with the correct runtime status but realized that doesn't work for client APIs like Terminate and Suspend. For those APIs, I tried updating the `Tags` in the relevant methods in `DurableClient` but was seeing different values for `Activity.Current` which meant the runtime status value wouldn't get updated correctly. Let me know if there's another way of approaching this that I haven't considered.

bachuv commented:
I was able to remove the additional state trace log by adding a `RuntimeStatus` field to the existing state trace log instead. I also updated the runtime status query.

```
traces
| where customDimensions.Category == "Host.Triggers.DurableTask"
| extend functionName = tostring(customDimensions["prop__functionName"])
| extend instanceId = tostring(customDimensions["prop__instanceId"])
| extend state = tostring(customDimensions["prop__state"])
| extend isReplay = tobool(tolower(customDimensions["prop__isReplay"]))
| extend hubName = tostring(tolower(customDimensions["prop__hubName"]))
| extend runtimeStatus = tostring((customDimensions["prop__runtimeStatus"]))
| where isReplay != true
| where state != "Awaited"
| where runtimeStatus != ""
| summarize arg_max(timestamp, *) by instanceId
| order by timestamp asc
| project timestamp, instanceId, runtimeStatus
```

<img width="203" alt="image" src="https://github.com/Azure/azure-functions-durable-extension/assets/15795068/468501cd-c81e-4959-bddd-a7ed0d78c374">

davidmrdavid commented:
Seems the FunctionsV2 tests are failing, @bachuv?

davidmrdavid commented on src/WebJobs.Extensions.DurableTask/WebJobs.Extensions.DurableTask.csproj:
```diff
@@ -109,6 +109,7 @@
     <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.DurableTask.Analyzers" Version="0.5.*" />
     <!-- Build-time dependencies -->
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.*" PrivateAssets="All" />
+    <PackageReference Include="Nunit3TestAdapter" Version="4.5.0" />
```
is this needed? Seems like an accidental addition

bachuv replied:
Nope, that was added on accident. I'll remove it!

davidmrdavid commented on test/Common/TestHelpers.cs:
```diff
@@ -483,8 +483,8 @@ public static Task DeleteTaskHubResources(string testName, bool enableExtendedSe
             {
-                // It is assumed that the 4th log message is a timer message.
-                timeStamp = GetTimerTimestamp(logMessages[3].FormattedMessage);
+                // It is assumed that the 5th log message is a timer message.
+                timeStamp = GetTimerTimestamp(logMessages[4].FormattedMessage);
```
I didn't catch why this suddenly became the 5th log (whereas before in was the 4th). Can you please clarify?

bachuv replied:
That was left over from the previous change when I was adding a new log for the runtime status change. For now, please ignore the changes in the test directory. I want to get feedback on the new approach (i.e. adding a new RuntimeStatus field to the existing state change log) before changing all of the tests since it's time consuming. Let me know if it's easier if I revert the changes in the test directory for now.

davidmrdavid replied:
Thanks for the clarification. Now that a reviews have been requested again, do we have some context on why the position of this message changed?

davidmrdavid reviewed (commented):
Left 2 small questions

davidmrdavid reviewed (commented):
With the exception of the `ResumingOrchestration` log, the code changes look good to me! It is certainly confusing that we log both a `State` and `RuntimeStatus` field. We'll need to tackle that _eventually_. Other than that, the overall change strategy makes sense to me. I think you should be good to update the test code

bachuv commented on test/Common/TestHelpers.cs:
The test changes are out of date and I'll change them once I get feedback on the new field and query. Please ignore the changes in this file. Thanks!

cgillum commented on src/WebJobs.Extensions.DurableTask/EndToEndTraceHelper.cs:
```diff
@@ -329,9 +329,9 @@ public void ExtensionWarningEvent(string hubName, string functionName, string in
-                instanceId, functionName, functionType, reason, FunctionState.Scheduled, hubName, LocalAppName,
-                LocalSlotName, ExtensionVersion, this.sequenceNumber++);
+                "{instanceId}: Resuming function '{functionName} ({functionType})'. Reason: {reason}. State: {state}. RuntimeStatus: {runtimeStatus}. HubName: {hubName}. AppName: {appName}. SlotName: {slotName}. ExtensionVersion: {extensionVersion}. SequenceNumber: {sequenceNumber}.",
+                instanceId, functionName, functionType, reason, FunctionState.Scheduled, OrchestrationRuntimeStatus.Pending, hubName,
```
+1, this should be `Running`

bachuv commented:
@davidmrdavid, I know the Netherite smoke test is expected to fail right now, but is the Python smoke test expected to fail?

davidmrdavid commented:
@bachuv: it isn't, but the error suggests the problem was a failure in downloading the docker image. I'm re-running the test now to see if it's transient.

cgillum reviewed (approved):
LGTM!
